### PR TITLE
Support behavior and event handling for Widget (#7435)

### PR DESCRIPTION
**Pull request by samdark**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #7435

### CODE REVIEW

File: framework/base/WidgetEvent.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * @author Petra Barus <petra.barus@gmail.com>
+ * @since 2.0.11
+ */
+class WidgetEvent extends Event {
+    
+    /**
+     * @var Widget the widget currently being executed
+     */
+    public $widget;
```
Comment (cebe) [inline framework/base/WidgetEvent.php]: isn't this duplicate with `$event->sender`?

File: framework/base/Widget.php [hunk 2]
```diff
@@ -220,4 +241,64 @@ public function getViewPath()
...
+     *
+     * @param mixed $result the widget return result.
+     * @return mixed the processed widget result.
+     * @since 2.0.11
+     */
+    public function afterRun($result)
+    {
+        $event = new WidgetEvent($this);
+        $event->result = $result;
+        $this->trigger(self::EVENT_BEFORE_RUN, $event);
```
Comment (cebe) [inline framework/base/Widget.php]: AFTER_RUN, not BEFORE_RUN.

**Comment by cebe** [comment]
An `EVENT_INIT` is also missing to cover the `begin()` `end()` case.

File: framework/base/WidgetEvent.php (hunk 1 above)
Comment (samdark) [inline framework/base/WidgetEvent.php]: It is...

File: framework/base/Widget.php (hunk 2 above)
Comment (samdark) [inline framework/base/Widget.php]: https://github.com/yiisoft/yii2/commit/767400da199f93994e9e181960d973744256828e

File: framework/base/WidgetEvent.php (hunk 1 above)
Comment (samdark) [inline framework/base/WidgetEvent.php]: https://github.com/yiisoft/yii2/commit/767400da199f93994e9e181960d973744256828e

**Pull request merged**